src/metabase/query_processor/middleware/visualization_settings.clj

```diff
@@ -0,0 +1,72 @@
+(ns metabase.query-processor.middleware.visualization-settings
+  (:require [medley.core :as m]
+            [metabase.models.card :refer [Card]]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.shared.models.visualization-settings :as mb.viz]
+            [toucan.db :as db]))
+
+(defn- normalize-field-settings
+  [id settings]
+  (mb.viz/db->norm-column-settings {(mb.viz/norm->db-column-ref {::mb.viz/field-id id}) settings}))
+
+(defn- update-card-viz-settings
+  "For each field, fetch its settings from the QP store, convert the settings into the normalized form
+  for visualization settings, and then merge in the card-level column settings."
+  [fields column-viz-settings]
+  (map (fn [[_ id _]]
+         (->> id
+              qp.store/field
+              :settings
+              (#(normalize-field-settings id %))
+              (#(m/deep-merge % (select-keys column-viz-settings [{::mb.viz/field-id id}])))))
+   fields))
+
+(defn- update-native-query-viz-settings
+  "For each column, select its viz settings from the column-viz-settings map by its name."
+  [columns column-viz-settings]
+  (map (fn [column] (-> column
+                        :name
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
```

Comment:
This sort of stuff would probably be more clear if you wrote it using `for`. Generally I think it's preferable to use higher-level looping constructs like `for` over `map` and threading macros, especially threading macros that work around the threading using inline `#()` functions